### ISSUE TITLE
Unknown type Jsonb when using Attributes API. 

### ISSUE THREAD

**Issue by jhirn** [rgeo/activerecord-postgis-adapter#267]
Hi. 

I'm having an issue when declaring a jsonb attribute using the attributes api and the postgis adapter. This works fine when using the `postgresql` adapter. 

If I have a model defined like this

```
class AttributeBug < ApplicationRecord
  attribute :foobar, :jsonb
end
```

I get the following error in rails console 

```
irb(main):003:0> AttributeBug.new
ArgumentError: Unknown type :jsonb
	from (irb):3
irb(main):004:0> AttributeBug.new
=> #<AttributeBug id: nil, created_at: nil, updated_at: nil>
```

Note the 2nd call doesn't fail, although the model does not have the `foobar` attribute defined on it. 

I've created a sample app here to recreate the problem. I tried to dig through this and the `pg` gem to see the difference in how the `ActiveRecord::Type`s are registered, but couldn't figure it out. I'll continue to search on my own but wanted to open an issue on it for now. 

https://github.com/jhirn/postgis_jsonb_attribute_bug/blob/master/README.md

**Comment by jhirn** [comment]
https://github.com/jhirn/postgis_jsonb_attribute_bug/blob/master/README.md

**Comment by teeparham** [comment]
Thanks for reporting this issue.

This adapter is a subclass of the Postgres adapter: `class PostGISAdapter < PostgreSQLAdapter`. The types are registered by the postgresql adapter here:
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

I'm not sure what the cause of the problem is, but maybe that will help.

In related news, I've noticed circular require warnings for active_record/postresql when running the test suite for rails 5.1 & haven't yet been able to track down the cause / solution there yet.

**Comment by jhirn** [comment]
Hey thanks for the tip (and the gem).  I'll try more debugging this afternoon.

**Comment by jhirn** [comment]
Hey @teeparham. 

I think I've found the issue, not sure how you'd like to go about resolving it in your adapter. The postgresql adapter registers these native types specific to the postgres adapter as such:

```
ActiveRecord::Type.register(:jsonb, OID::Jsonb, adapter: :posgresql)
```

[The source for that is here](https://github.com/rails/rails/blob/3f1695bb9c008b7cb1840e09e640f3ec0c59a564/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L813)

I've found by registering these with the adapter set to postgis, the problem goes away. 

```
module ActiveRecord
  module ConnectionAdapters
    ActiveRecord::Type.register(:jsonb, PostgreSQL::OID::Jsonb, adapter: :postgis)
    #and so on.....
  end
end
```

I guess you could just duplicate the registrations in the PostGIS connection adapter. It's unlikely they'll suffer from a ton of churn. It would be a lot nicer maintenance wise if you could read metadata from the PostgreSQL registry and repeat them for PostGIS adapter. Not exactly sure how to do that but going to poke around. This is deeper than I've had to go before in AR ;)

If I find a nice solution I'll issue a PR for you

**Comment by jhirn** [comment]
I couldn't find a way to read decent way to read the registry and get the collection of types initialized in `PostgreSQLAdapter` registered with PostGIS. For now I've monkey patched it in my project. Here's my fork with the commit. Happy to make a PR if it's suitable. Please note for 4.0-stable you have to omit `legacy_point` type. I believe it's not available in AR 5.0. 

https://github.com/jhirn/activerecord-postgis-adapter/commit/04a8cf09c433af5c10260640c50a1b0ff91a507b

**Comment by teeparham** [comment]
I still don't have a solution but here's what I found/confirmed:

The types are registered to the "postgresql" adapter (even though we're using "postgis"). New spatial types are not registered. Ideally the upstream type registration in the PostgreSQLAdapater would work for subclassed adapters, & then the PostGIS adapter could add the new spatial types.

```ruby
# ActiveRecord 5.0
ActiveRecord::Type.registry.send(:registrations).map{|r| [r.send(:adapter), r.send(:name)].join(" : ")}
[
  [ 0] " : big_integer",
  [ 1] " : binary",
  [ 2] " : boolean",
  [ 3] " : date",
  [ 4] " : datetime",
  [ 5] " : decimal",
  [ 6] " : float",
  [ 7] " : integer",
  [ 8] " : string",
  [ 9] " : text",
  [10] " : time",
  [11] "postgresql : ",
  [12] "postgresql : ",
  [13] "postgresql : bit",
  [14] "postgresql : bit_varying",
  [15] "postgresql : binary",
  [16] "postgresql : cidr",
  [17] "postgresql : datetime",
  [18] "postgresql : decimal",
  [19] "postgresql : enum",
  [20] "postgresql : hstore",
  [21] "postgresql : inet",
  [22] "postgresql : json",
  [23] "postgresql : jsonb",
  [24] "postgresql : money",
  [25] "postgresql : point",
  [26] "postgresql : legacy_point",
  [27] "postgresql : uuid",
  [28] "postgresql : vector",
  [29] "postgresql : xml"
```

It seems possible to alter the type registry when the PostGIS adapter is loaded (change "postgresql" to "postgis"). For example, this hackery works:

```ruby
ActiveRecord::Type.registry.send(:registrations).last.instance_variable_set("@adapter", "postgis")
```

**Comment by jhirn** [comment]
Yeah. For now I’ve just copied the registrations from the Postgres get rather than try to do it dynamically. I’m loading them via an initializer. 

FWIW, the set is so infrequently changed it may be worth just registering these types manually and doing a point release when something new comes along (or goes away).

**Comment by keithdoggett** [comment]
With #334, the adapter will now check for types registered under the `:postgis` or `:postgresql` adapters.

**Comment by jhirn** [comment]
yay!

**Comment by SampsonCrowley** [comment]
@keithdoggett has there been a new release since that was merged? I just ran into this in combination with another gem (that relies on type lookups) not being able to find type `uuid` (full error: `ArgumentError (Unknown type :uuid)`) that instantly goes away when switched to `:postgresql` instead of `:postgis`

**Comment by SampsonCrowley** [comment]
Even just throwing the types registration into an override module fixes it for me though, so the fix is definitely functional! just either waiting on or missed the next release

**Comment by keithdoggett** [comment]
@SampsonCrowley this has been released in 7.1.0 and is only in the master branch currently.
